### PR TITLE
provide API for saving documents

### PR DESCRIPTION
**Pull request by kevinushey**
This PR provides `.rs.api.documentSave()` as an interface for saving (file-backed) files within RStudio. The main intention is to allow for packages (e.g. `devtools`) to save open documents before performing some action.

Because we use a `WaitForMethod` to accomplish this (request a document save and then wait for that to finish), it's prudent that the completion event is fired regardless of whether we save successfully or not. (Otherwise, we risk getting the `rsession` stuck waiting for an event that will never come)

@jmcphers, could you review? I wanted to avoid touching the plumbing around saves; let me know if you think there's a cleaner way forward. (see the logic within `onRequestDocumentSave()`)

See also: https://github.com/rstudio/rstudioapi/pull/49

### CODE REVIEW

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java [hunk 1]
```diff
@@ -4230,6 +4234,87 @@ public void onOpenProfileEvent(OpenProfileEvent event)
...
+      // the current save code is wired up in such a way that it's difficult
+      // to distinguish a success from failure, so we just make sure the
+      // server receives a response after 5s (defaulting to failure)
+      final Mutable<Boolean> savedSuccessfully = new Mutable<Boolean>(false);
+      final Timer completedTimer = new Timer()
+      {
+         @Override
+         public void run()
+         {
+            server_.requestDocumentSaveCompleted(
```
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java]: How does this work with source windows? Seems like as written each window will wind up calling this, which is probably not what you want. If it's not clear to you how to wire this up ping me (the problem is already largely solved in the existing quit logic).

**Comment by kevinushey** [comment]
I think this PR is ready for re-review. We now hook into the same logic provided by the `saveAll()` APIs, but augment it with a custom predicate function that can be used to further filter the selection of targets upon which the save action is performed.

I'm a little bit worried since the changes here imply that we're now using GWT objects in a cross-window fashion, but (based on some initial testing) this seems to work in practice. I'll spend a bit time tomorrow seeing whether we can eschew that just in the interest of safety / correctness.

**Comment by kevinushey** [comment]
I tried working my way through explicit use of the JavaScriptSerializer, but unfortunately had no luck there.

Given that this does seem to work, perhaps the PR as-is is fine? (Presumedly failure in this sort of cross-window object usage would result from errant usages of `instanceof`, but it seems like we're not bumping into that). Or are there other concerns that I may be missing?

**Comment by jmcphers** [comment]
For some background: some Java code patterns can result in implicit dynamic casts, which will be expressed explicitly in the generated Javascript and can throw an exception at runtime. These dynamic casts rely on JavaScript object identities so don't work cross-window. Therefore it's pretty dangerous to pass objects cross-window because they are very volatile in the target window, often resulting in exceptions and bugs that are hard to trace when they are first dynamic-cast to their expected type. Even if the existing code doesn't trip any wires it may create some fragility. 

I realize however that it may be the only good way to solve the problem so will take a more thorough look!

File: src/gwt/src/org/rstudio/studio/client/common/JSONArrayBuilder.java [hunk 2]
```diff
@@ -61,10 +61,15 @@ public final JSONArrayBuilder add(double value)
    
    public final JSONArrayBuilder add(JsArrayString value)
    {
+      if (value == null)
+         return appendNull();
+      
       JSONArray array = new JSONArray();
-      for (int i = 0, n = value.length(); i < n; i++)
-         array.set(i, fromString(value.get(i)));
+      if (value != null)
```
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/common/JSONArrayBuilder.java]: Is this test superfluous? (looks like we'd have already have bailed if value was null)

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/SourceWindowManager.java [hunk 3]
```diff
@@ -944,9 +947,11 @@ private final native void saveWithPrompt(WindowEx satellite,
...
    }-*/;
 
-   private final native void saveAllUnsaved(WindowEx satellite, 
-         Command onCompleted) /*-{
-      satellite.rstudioSaveAllUnsaved(onCompleted);
+   private final native void saveUnsavedDocuments(WindowEx satellite,
+                                                  EditingTargetSavePredicate predicate,
+                                                  Command onCompleted)
+   /*-{
+      satellite.rstudioSaveUnsavedDocuments(predicate, onCompleted);
```
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/SourceWindowManager.java]: It seems like the primary purpose of the predicate is to filter by ID, so if you wanted to mitigate any cross-window issues down the road you could pass POD like a JsArrayString of IDs instead. However I'll let you make the call on that based on your perception of the risk of the predicate evolving to become more complex.

**Comment by kevinushey** [comment]
Let me take one more stab at this moving from the Java predicate to a plain old JsArray; if I hit some roadblocks then I'll merge the PR as is. Thanks for reviewing + the insight!

File: src/gwt/src/org/rstudio/studio/client/common/JSONArrayBuilder.java (hunk 2 above)
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/common/JSONArrayBuilder.java]: Indeed it is -- thanks for catching that!

**Comment by kevinushey** [comment]
Now using `Set<String>` for the user-facing APIs, and marshaling to and from `JsArrayString` when crossing the window boundary. Thanks!

**Comment by jmcphers** [comment]
Nice! Thanks for taking care of that.

**Pull request merged**